tests/gold_tests/tls/tls_tunnel.test.py

```diff
@@ -96,19 +96,19 @@
     "- fqdn: bob.*.com",
     "  tunnel_route: localhost:{0}".format(server_foo.Variables.SSL_Port),
     "- fqdn: '*.match.com'",
-    "  tunnel_route: $1.testmatch:{0}".format(server_foo.Variables.SSL_Port),
+    '''  tunnel_route: "{{1}}.testmatch:{0}"'''.format(server_foo.Variables.SSL_Port),
     "- fqdn: '*.ok.*.com'",
-    "  tunnel_route: $2.example.$1:{0}".format(server_foo.Variables.SSL_Port),
+    '''  tunnel_route: "{{2}}.example.{{1}}:{0}"'''.format(server_foo.Variables.SSL_Port),
     "- fqdn: ''",  # No SNI sent
     "  tunnel_route: localhost:{0}".format(server_bar.Variables.SSL_Port),
     "- fqdn: 'incoming.port.com'",
-    "  tunnel_route: backend.incoming.port.com:{inbound_local_port}",
+    '''  tunnel_route: "backend.incoming.port.com:{inbound_local_port}"''',
     "- fqdn: 'proxy.protocol.port.com'",
-    "  tunnel_route: backend.proxy.protocol.port.com:{proxy_protocol_port}",
+    '''  tunnel_route: "backend.proxy.protocol.port.com:{proxy_protocol_port}"''',
     "- fqdn: '*.*.incoming.port.com'",
-    "  tunnel_route: backend.$1.$2.incoming.port.com:{inbound_local_port}",
+    '''  tunnel_route: "backend.{1}.{2}.incoming.port.com:{inbound_local_port}"''',
```

Comment:
Is it valid for  `tunnel_route` url to have the `{}` characters?  I am not sure if  we support url that contains the `{}` at all in ATS since they are considered unsafe and need to be percent-encoded. But if so, there would be issues for:

- When `{abc}` is in the url but `abc` is an invalid name for bwf so would become `~abc~`
- When `{3}` is specified in the url but only two match groups are available.

Reply:
For the first point, the braces are removed by the substitution, just as is done currently. It would be expected that a correctly formatted string would not have any braces in the final result.

For the error cases, I pondered that. What happens current if "$3" is present but there are insufficient match groups? I would need to make a tweak to libswoc to provide access to the parsed format, after which the logic could check for invalid substitutions for names and match groups.

Reply:
Another thing to consider is, for backwards compatibility, do a pre-pass that converts "[$]([0-9]+)" to "{\1}".

Reply:
> What happens current if "$3" is present but there are insufficient match groups?

Currently the `$3` would be left as is for this case.